storage/block.go

```diff
@@ -124,27 +122,26 @@ func (b *dbBlock) close() {
 		return
 	}
 
-	// Otherwise, we need to close the stream.
-	if stream := b.stream; stream != nil {
-		b.ctx.RegisterCloser(func() {
-			if bytesPool := b.opts.GetBytesPool(); bytesPool != nil {
-				segment := stream.Segment()
-				stream.Reset(m3db.Segment{})
-				bytesPool.Put(segment.Head)
-				bytesPool.Put(segment.Tail)
-			}
-			stream.Close()
-		})
-	}
+	// Otherwise, we need to return bytes to the bytes pool.
+	segment := b.segment
+	bytesPool := b.opts.GetBytesPool()
+	b.ctx.RegisterCloser(func() {
+		// NB(xichen): we do not return segment tail here because the tails
+		// are precomputed in the encoding options and therefore potentially
+		// reused across different encoders.
+		if segment.Head != nil {
+			bytesPool.Put(segment.Head)
+		}
+	})
 }
 
 // Reset resets the block start time and the encoder.
 func (b *dbBlock) Reset(startTime time.Time, encoder m3db.Encoder) {
 	b.close()
 	b.start = startTime
 	b.encoder = encoder
-	b.stream = nil
-	b.ctx = b.ctxAlloc()
+	b.segment = m3db.Segment{}
+	b.ctx = b.opts.GetContextPool().Get()
```

Comment:
To make sure we release the old context we should `b.ctx.Close()` before `b.ctx = b.opts.GetContextPool().Get()`